src/librustc_mir/interpret/intern.rs

```diff
@@ -430,3 +431,25 @@ pub fn intern_const_alloc_recursive<M: CompileTimeMachine<'mir, 'tcx>>(
         }
     }
 }
+
+impl<'mir, 'tcx: 'mir, M: super::intern::CompileTimeMachine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
+    /// A helper function that allocates memory for the layout given and gives you access to mutate
+    /// it. Once your own mutation code is done, the backing `Allocation` is removed from the
```

Comment:
Ok, so when we use this in const-prop, doesn't the `Memory` get dropped when the pass finishes and drops its `InterpCx`? Doesn't that leave this `AllocId` "dangling" by the time we reach codegen? 

(I realize I'm wrong because this bootstraps and passes all tests, but I'd like to understand which part I'm wrong about)

Reply:
Yes, that `AllocId` is dangling, but there's nothing with that `AllocId` anymore unless the caller of `intern_with_temp_alloc` copies it out of the `f` closure while it's being run. Note that it's dangling even before the `InterpCx` and `Memory` are dropped, as the `self.memory.alloc_map.remove` takes it out of there without even inserting anything into the `dead_alloc` table. This means any mis-use will give you an ICE.

What this function returns is a `&'tcx Allocation`, which doesn't need an `AllocId` that points to it. If a user of `intern_with_temp_alloc` wants an `AllocId` for the Allocation, they can invoke https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/struct.TyCtxt.html#method.create_memory_alloc which will give them a new `AllocId`

Reply:
Got it. That makes perfect sense. Thanks for the explanation!